### DISCUSSION TITLE
dgemm_ bug? this is only triggered when N=200-201 M=6-7

### DISCUSSION THREAD

**Discussion by Xie-ry** [OpenMathLib/OpenBLAS#4506] [Q&A]
I found a strange thing about the dgemm_ function.
This is pseudocode, the full code is given later.
```
#include "OpenBLAS/include/f77blas.h"
double alpha = 1.0;
double beta  = 0.0;

double* bb = GetRandomMartix(N, N); // malloc
double* cc = GetRandomMartix(M, N); //

// test 1
double* aa = GetRandomMartix(N, M); // a TRANS in gemm
dgemm_("T", "N", &M, &N, &N, &alpha, aa, &N, bb, &N, &beta, cc, &M);

// test 2
double* aa = GetRandomMartix(M, N); // a TRANS in create
dgemm_("N", "N", &M, &N, &N, &alpha, aa, &M, bb, &N, &beta, cc, &M);
```

strange results：
```
N = 200 M = 7  and  for loop 100 times
test1 need 7.04418ms
test2 need 4.03527ms

N = 200 M = 6  and  for loop 100 times
test1 need 2.39609ms
test2 need 4.02413ms

N = 201 M = 6  and  for loop 100 times
test1 need 7.50546ms
test2 need 4.07308ms
```
Can anyone explain this phenomenon? Just by adding a row or a column, the time increases several times.

env:
```
OpenBLAS 3.26 default compilation
gcc 9.1 or gcc 12.3.0
Intel(R) Xeon(R) Gold 6248R CPU
Single Thread, OMP_NUM_THREADS=1 BLAS_NUM_THREADS=1
```

**Comment by martin-frbg** [comment]
That is probably the crossover point where OpenBLAS "decides" the problem is large enough to switch to multithreading (and the decision is made too early in case 1). Need to look into this...

**Reply by Xie-ry** [reply]
Env is added. The test is performed only in single-thread mode. And use taskset -c 2 （one cpu core）.

**Reply by martin-frbg** [reply]
thank you. maybe something weird with loop unrolling, or at that point things simply do no longer fit in the cpu cache

**Comment by Xie-ry** [comment]
[mat_test_gemm.txt](https://github.com/OpenMathLib/OpenBLAS/files/14321949/mat_test_gemm.txt)

This is the complete test code that can be run.

**Comment by Xie-ry** [comment]
@martin-frbg Maybe it's the matrix transpose that changes the return of _GEMM_SMALL_MATRIX_PERMIT_.
```
// interface/gemm.c:502
// if(GEMM_SMALL_MATRIX_PERMIT(transa, transb, args.m, args.n, args.k, *(FLOAT *)(args.alpha), *(FLOAT *)(args.beta))) {
// I modify this line, replacing it with the following line:
if(0) {

// Then I found that the timing of the 2 tests became close.
N = 200 M = 7  and  for loop 100 times
test1 need 6.1039ms
test2 need 6.26614ms

N = 200 M = 6  and  for loop 100 times
test1 need 5.81852ms
test2 need 7.05903ms

N = 201 M = 6  and  for loop 100 times
test1 need 5.95921ms
test2 need 6.49428ms

// Trace _GEMM_SMALL_MATRIX_PERMIT_ source
// GEMM_SMALL_MATRIX_PERMIT -> kernel\x86_64\dgemm_small_kernel_permit_skylakex.c
if (M * N > 1200 || K < 32)
    return 0;
```
Am I doing the right thing? My knowledge of matrices is weak. I try to explain it logically from the code.
if no use _GEMM_SMALL_MATRIX_PERMIT_, performance seems to be down
